crates/matrix-sdk-base/src/calls.rs

```diff
@@ -0,0 +1,129 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Things related to the room calls (also known as MatrixRTC sessions).
+
+use std::{collections::BTreeMap, time::Duration};
+
+use ruma::{MilliSecondsSinceUnixEpoch, OwnedDeviceId, OwnedUserId};
+
+/// Represents a room call (MatrixRTC session). A call consists of a set of
+/// members who participate in a call. A call is active as long as there is at
+/// least one member (participant) in it.
+#[derive(Debug)]
+pub struct RoomCall {
+    /// The members (participants) of the call as of the moment when the call
+    /// was fetched.
+    pub(crate) members: BTreeMap<CallMemberIdentifier, CallMemberData>,
+}
+
+impl RoomCall {
+    /// Returns the time when the call was started, i.e. when the first member
+    /// joined.
+    ///
+    /// Returns `None` if there are no members in the call (if the call is
+    /// empty/ended).
+    pub fn start_time(&self) -> Option<MilliSecondsSinceUnixEpoch> {
+        // TODO: We must take into an account that a call could have been started by a
+        // participant who is no longer a member of a call.
+        self.members.values().map(|member| member.member_since).min()
+    }
+
+    /// Non-expired members of a call.
+    pub fn active_members(&self) -> impl Iterator<Item = &CallMemberIdentifier> {
+        self.members.iter().filter(|(_, member)| !member.expired()).map(|(id, _)| id)
+    }
+}
+
+/// A unique identifier of a call member (participant).
+/// Each call member is identified by their user ID and device ID.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct CallMemberIdentifier {
+    /// The user ID of the call member.
+    pub user_id: OwnedUserId,
+    /// The device ID of the call member.
+    pub device_id: OwnedDeviceId,
+}
+
+/// Data about a call member (participant).
+#[derive(Clone, Debug)]
+pub struct CallMemberData {
+    /// The time when the member joined the call.
+    pub member_since: MilliSecondsSinceUnixEpoch,
+    /// The duration for which the member is valid starting from
+    /// [`Self::member_since`].
+    pub valid_for: Duration,
+}
+
+impl CallMemberData {
+    /// Returns `true` if the member is still valid, i.e. if the member joined
+    /// and its membership has not expired yet. Returns `false` otherwise.
+    pub fn expired(&self) -> bool {
+        let expires_at = self.member_since.to_system_time().map(|start| start + self.valid_for);
+        let now = MilliSecondsSinceUnixEpoch::now().to_system_time();
+        expires_at.zip(now).map(|(expires_at, now)| expires_at < now).unwrap_or(true)
+    }
+}
```

Comment:
We can't use `Membership`, because the `Membership` is a structure that contains everything related to a participant asserting to be part of a call/session, i.e. it contains fields like a `call_id`, an optional `created_at` (that by this point must not be optional anymore) and the call type (the field that at this point is part of a call, not a member itself), i.e. the call is an aggregated state of all relevant memberships.

Essentially, `CallMemberData` is the relevant/calculated data from the `Membership` of a particular call member (call participant).